engine/src/main/java/org/terasology/rendering/world/dag/ShadowMapNode.java

```diff
@@ -0,0 +1,171 @@
+        // TODO: shader and fbo shall be initialised here
+        // TODO: using ShaderParametersShadowMap here, instead of storing in ShaderManagerLwjgl
+    }
+
+    @Override
+    public void preRender(float update) {
+        // positionShadowMapCamera()
+
+        // Shadows are rendered around the player so...
+        Vector3f lightPosition = new Vector3f(playerCamera.getPosition().x, 0.0f, playerCamera.getPosition().z);
+
+        // Project the shadowMapCamera position to light space and make sure it is only moved in texel steps (avoids flickering when moving the shadowMapCamera)
+        float texelSize = 1.0f / renderingConfig.getShadowMapResolution();
+        texelSize *= 2.0f;
+
+        camera.getViewProjectionMatrix().transformPoint(lightPosition);
+        lightPosition.set(TeraMath.fastFloor(lightPosition.x / texelSize) * texelSize, 0.0f, TeraMath.fastFloor(lightPosition.z / texelSize) * texelSize);
+        camera.getInverseViewProjectionMatrix().transformPoint(lightPosition);
+
+        // ... we position our new shadowMapCamera at the position of the player and move it
+        // quite a bit into the direction of the sun (our main light).
+
+        // Make sure the sun does not move too often since it causes massive shadow flickering (from hell to the max)!
+        float stepSize = 50f;
+        Vector3f sunDirection = backdropProvider.getQuantizedSunDirection(stepSize);
+
+        Vector3f sunPosition = new Vector3f(sunDirection);
+        sunPosition.scale(256.0f + 64.0f);
+        lightPosition.add(sunPosition);
+
+        camera.getPosition().set(lightPosition);
+
+        // and adjust it to look from the sun direction into the direction of our player
+        Vector3f negSunDirection = new Vector3f(sunDirection);
+        negSunDirection.scale(-1.0f);
+
+        camera.getViewingDirection().set(negSunDirection);
+
+        camera.update(update);
+
+    }
+
+
+    @Override
+    public void render() {
+        // TODO: removing this assignment
+        fbo = state.buffers.sceneShadowMap; // added here since graphic settings can be modified and a new fbo can be generated.
+
+        // preRenderSetupSceneShadowMap
+        fbo.bind();
+        glViewport(0, 0, fbo.width(), fbo.height());
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+        GL11.glDisable(GL_CULL_FACE);
+
+        // render
+        camera.lookThrough();
+
+        shader.enable();
+        // FIXME: storing chuncksOpaqueShadow or a mechanism for requesting a chunk queue for nodes which calls renderChunks method?
+        worldRenderer.renderChunks(worldRenderer.renderQueues.chunksOpaqueShadow, ChunkMesh.RenderPhase.OPAQUE, camera, WorldRendererImpl.ChunkRenderMode.SHADOW_MAP);
+        playerCamera.lookThrough(); //FIXME: not strictly needed: just defensive programming here.
+
+
+        // postRenderCleanupSceneShadowMap
+        GL11.glEnable(GL_CULL_FACE);
+        glBindFramebufferEXT(GL_FRAMEBUFFER_EXT, 0); // bindDisplay()
```

Comment:
Similarly to above, I'd rather get this back to bindDisplay(). Also, the postRenderCleanupSceneShadowMap() method had an additional line: setViewportToWholeDisplay(); It's probably redundant but for the time being we should leave it there. Perhaps the way to do this is to keep GraphicState until it is reduced to a skeleton (without all the preRenderSetup/postRenderCleanup methods) and then rename it to something more meaningful, i.e. OpenGLHelper class or OpenGLUtils or something along those lines.